File: Week4 (Sort, Search, Tree-1)/Sorting/lbh_selection_insertion_sort.py
def selection_sort(arr):
    for i in range(len(arr)):
        min_idx = i
        for j in range(i+1, len(arr)):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]

def insertion_sort(arr):
    for i in range(1, len(arr)):  # Iterate from the second element to the end of the array
        key = arr[i]  # Take the current element (element to be inserted in the sorted portion)
        j = i - 1  # Start comparing with the element just before the current element

        # Move elements of arr[0..i-1], that are greater than key, to one position ahead of their current position
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]  # Shift the element one position to the right
            j -= 1  # Move to the previous element

        arr[j + 1] = key  # Insert the key in its correct position

# Driver code
arr = [64, 25, 12, 22, 11]
insertion_sort(arr)
print("Sorted array:", arr)

File: Week4 (Sort, Search, Tree1)/Sorting/lbh_selection_insertion_sort.cpp
#include <iostream>
using namespace std;

// Function to swap two elements
void swap(int &xp, int &yp) {
    int temp = xp;
    xp = yp;
    yp = temp;
}

// Selection Sort function
void selectionSort(int arr[], int n) {
    int i, j, min_idx;
    for (i = 0; i < n-1; i++) {
        // Find the minimum element in the unsorted array
        min_idx = i;
        for (j = i+1; j < n; j++)
            if (arr[j] < arr[min_idx])
                min_idx = j;
        
        // Swap the found minimum element with the first element
        swap(arr[min_idx], arr[i]);
    }
}

// Insertion Sort function
void insertionSort(int arr[], int n) {
    int i, key, j;
    for (i = 1; i < n; i++) {
        key = arr[i];
        j = i - 1;

        // Move elements of arr[0..i-1], that are greater than key, to one position ahead of their current position
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }
}

// Function to print an array
void printArray(int arr[], int size) {
    for (int i = 0; i < size; i++)
        cout << arr[i] << " ";
    cout << endl;
}

// Driver code
int main() {
    int arr[] = {64, 25, 12, 22, 11};
    int n = sizeof(arr)/sizeof(arr[0]);
    insertionSort(arr, n);
    cout << "Sorted array: \n";
    printArray(arr, n);
    return 0;
}